include/rmm/mr/device/system_memory_resource.hpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace rmm::mr {
+
+namespace detail {
+/**
+ * @brief Check if system allocated memory (SAM) is supported on the specified device.
+ *
+ * @param device_id The device to check
+ * @return true if SAM is supported on the device, false otherwise
+ */
+static bool is_system_memory_supported(cuda_device_id device_id)
+{
+  int pageableMemoryAccess;
+  RMM_CUDA_TRY(cudaDeviceGetAttribute(
+    &pageableMemoryAccess, cudaDevAttrPageableMemoryAccess, device_id.value()));
+  return pageableMemoryAccess == 1;
+}
+}  // namespace detail
+
+/**
+ * @addtogroup device_memory_resources
+ * @{
+ * @file
+ */
+/**
+ * @brief `device_memory_resource` derived class that uses malloc/free for allocation/deallocation.
+ *
+ * There are two flavors of hardware/software environments that support accessing system allocated
+ * memory (SAM) from the GPU: HMM and ATS.
+ *
+ * Heterogeneous Memory Management (HMM) is a software-based solution for PCIe-connected GPUs on
+ * x86 systems. Requirements:
+ *   - NVIDIA CUDA 12.2 with the open-source r535_00 driver or newer.
+ *   - A sufficiently recent Linux kernel: 6.1.24+, 6.2.11+, or 6.3+.
+ *   - A GPU with one of the following supported architectures: NVIDIA Turing, NVIDIA Ampere,
+ *     NVIDIA Ada Lovelace, NVIDIA Hopper, or newer.
+ *   - A 64-bit x86 CPU.
+ *
+ *  For more information, see
+ *  https://developer.nvidia.com/blog/simplifying-gpu-application-development-with-heterogeneous-memory-management/.
+ *
+ *  Address Translation Services (ATS) is a hardware/software solution for the Grace Hopper
+ *  Superchip that uses the NVLink Chip-2-Chip (C2C) interconnect to provide coherent memory. For
+ *  more information, see
+ *  https://developer.nvidia.com/blog/nvidia-grace-hopper-superchip-architecture-in-depth/.
+ */
+class system_memory_resource final : public device_memory_resource {
```

Comment:
We have a use case where we want a system_memory_resource that is not limited to systems that support paging. So we don't want to limit it to `is_system_memory_supported(device)`. Instead, we want to limit the `device_accessible` property to be true on systems where the memory is device accessible. I'm currently checking whether the device_accessible property can be used as a dynamic property. I don't think so. :(

Reply:
I guess technically you could have multiple GPUs on a system and only some of them support SAM (e.g. a Turing and a Pascal), so the dynamic property also needs a device.

Reply:
@harrism can this be merged now or do you still want to get the device_accessible property resolved first?

Reply:
Still waiting, have pinged @jrhemstad for help. I think since MRs are implicitly tied to the current device, it can query the current device, doesn't need to take a parameter.

Reply:
`device_accessible` is not a dynamic property.

For this kind of functionality, we'd need to add a new property like `maybe_device_accessible` where `get_property(mr, cuda::maybe_device_accessible)` returns a `bool` for if it is device accessible or not.

Reply:
Can we merge this now? We can always enhance it when we have a dynamic `device_accessible`.

Reply:
@jrhemstad that would make this MR unusable with a resource_ref that expects `device_accessible`.

Reply:
@rongou Perhaps. I'm troubled that something called a `system_memory_resource` is only usable on systems that support paging memory to device.

Reply:
It's a `system_memory_resource` under the `device` namespace, so it's probably reasonable to assume it only works on environments that support accessing system allocated memory from a device.

Reply:
There is no `device` namespace, and `device_memory_resource` and `host_memory_resource` base classes are planned for removal.